vibrato/src/trainer.rs

```diff
@@ -403,11 +396,254 @@ impl Trainer {
+    pub fn write_used_features<L, R>(&mut self, left_wtr: L, right_wtr: R) -> Result<()>
+    where
+        L: Write,
+        R: Write,
+    {
+        if self.merged_model.is_none() {
+            self.merged_model = Some(self.data.raw_model.merge()?);
+        }
+        let merged_model = self.merged_model.as_ref().unwrap();
+
+        // left
+        let mut left_wtr = BufWriter::new(left_wtr);
+        let mut left_features: Vec<_> = self
+            .data
+            .config
+            .feature_extractor
+            .left_feature_ids()
+            .iter()
+            .collect();
+        left_features.sort_unstable_by_key(|(_, v)| **v);
+        let feature_list = &merged_model.left_conn_to_right_feats;
+        for (conn_id, feat_ids) in feature_list[..feature_list.len() - 1].iter().enumerate() {
+            write!(&mut left_wtr, "{}", conn_id + 1)?;
+            for (i, feat_id) in feat_ids.iter().enumerate() {
+                if let Some(feat_id) = feat_id {
+                    let feat_str = &left_features[usize::from_u32(feat_id.get()) - 1].0;
+                    write!(&mut left_wtr, " {i}:{feat_str}")?;
+                }
+            }
+            writeln!(&mut left_wtr)?;
+        }
+
+        // right
+        let mut right_wtr = BufWriter::new(right_wtr);
+        let mut right_features: Vec<_> = self
+            .data
+            .config
+            .feature_extractor
+            .right_feature_ids()
+            .iter()
+            .collect();
+        right_features.sort_unstable_by_key(|(_, v)| **v);
+        let feature_list = &merged_model.right_conn_to_left_feats;
+        for (conn_id, feat_ids) in feature_list[..feature_list.len() - 1].iter().enumerate() {
+            write!(&mut right_wtr, "{}", conn_id + 1)?;
+            for (i, feat_id) in feat_ids.iter().enumerate() {
+                if let Some(feat_id) = feat_id {
+                    let feat_str = &right_features[usize::from_u32(feat_id.get()) - 1].0;
+                    write!(&mut right_wtr, " {i}:{feat_str}")?;
+                }
+            }
+            writeln!(&mut right_wtr)?;
+        }
+        Ok(())
+    }
+
+    /// Write the dictionary.
+    ///
+    /// # Arguments
+    ///
+    /// * `lexicon_wtr` - Write sink targetting `lex.csv`.
+    /// * `connector_wtr` - Write sink targetting `matrix.def`.
+    /// * `unk_handler_wtr` - Write sink targetting `unk.def`.
+    /// * `user_lexicon_wtr` - Write sink targetting `user.csv`. If no user-defined lexicon file is
+    ///                        specified, this function generates an empty file.
+    ///
+    /// # Errors
+    ///
+    /// [`VibratoError`](crate::errors::VibratoError) is returned when:
+    ///
+    /// - the merging weights fails, or
+    /// - the writing fails.
+    pub fn write_dictionary<L, C, U, S>(
+        &mut self,
+        lexicon_wtr: L,
+        connector_wtr: C,
+        unk_handler_wtr: U,
+        user_lexicon_wtr: S,
```

Comment:
I guess that this argument is a bit unkind to those who do not need a user dictionary because a writer instance has to be always created.
How about using `Option`?

Reply:
If the argument is `Option<S>`, it is inconvenient because the type `S` must be specified explicitly.

A separate function to output the user dictionary would solve this problem, but is is not good design because the scaling factor is currently calculated from the maximum value of both the system dictionary and the user dictionary.

If the scaling factors are calculated only from the system dictionary, the functions can be separated, but the maximum weight value of 32767 may be exceeded depending on the features of the user dictionary, so I want to use both the system dictionary and the user dictionary for the calculation of the scaling factors.

Reply:
I understand and agree with maintaining the argument.
I added a new review for the comment.